### ISSUE TITLE
[FEATURE]

### ISSUE THREAD

**Issue by pelumitegbe** [oss-slu/mithridatium#15]
## Summary

Create a single source of truth for image preprocessing before going into the model (size, mean/std, etc.).
We’ll store these settings in a small sidecar JSON file that sits next to the model (e.g., models/resnet18_bd.json).
Code will read that file and build the right transforms for all components (evaluator, MMBD, CLI).

##Acceptance Criteria

 - [ ]PreprocessConfig dataclass exists in mithridatium/io.py (fields documented with comments).

 - [ ] load_preprocess_config(model_path) reads models/<name>.json and returns a valid config.

 - [ ] build_dataloader(..., pp) constructs a working dataloader using the config (CIFAR-10 supported).

 - [ ] A simple test proves the JSON → PreprocessConfig → dataloader path works (iterate 1 batch).

- [ ] scripts/check_evaluator.py (or equivalent) uses the new loader so everyone follows the same preprocessing.

## Tasks that need to completed for this feature

1) Add a config dataclass (code-side)

File: mithridatium/io.py

Add:

`
from dataclasses import dataclass
from typing import Tuple, List

@dataclass
class PreprocessConfig:
    input_size: Tuple[int, int]      # (H, W)
    channels_first: bool             # True = NCHW, False = NHWC

##True → NCHW (batch, channels, height, width) — common in PyTorch.
##False → NHWC — common in some TF models or when exporting.

    value_range: Tuple[float, float] # pixel range after transforms, e.g., (0.0, 1.0)
    mean: Tuple[float, float, float] # per-channel mean (R, G, B)
    std:  Tuple[float, float, float] # per-channel std  (R, G, B)
"""
Example (CIFAR-10):

mean=(0.4914, 0.4822, 0.4465)

std =(0.2023, 0.1994, 0.2010)
These must match what the model was trained with, or accuracy (and detectors) can degrade.
"""

    ops:  List[str]                  # high-level ops, e.g., ["resize:32", "center_crop:32"]
`

2) Load config from a sidecar JSON

File: mithridatium/modelcard.py

Add:
`
import json
from pathlib import Path
from .io import PreprocessConfig

def load_preprocess_config(model_path: str) -> PreprocessConfig:
    # If model is models/resnet18_bd.pth → expect models/resnet18_bd.json
    card_path = Path(model_path).with_suffix(".json")
    data = json.loads(card_path.read_text())
    pp = data["preprocess"]
    return PreprocessConfig(
        input_size=tuple(pp["input_size"]),
        channels_first=pp.get("channels_first", True),
        value_range=tuple(pp.get("value_range", (0.0, 1.0))),
        mean=tuple(pp["mean"]),
        std=tuple(pp["std"]),
        ops=list(pp.get("ops", [])),
    )
`
Goal: Read a small JSON file that sits next to your model (the “sidecar”) and turn it into a PreprocessConfig.

If your model is models/resnet18_bd.pth, we expect a partner file models/resnet18_bd.json.

Code changes the file extension from .pth → .json and opens that JSON.

It reads the "preprocess" section and fills a PreprocessConfig dataclass (fields like mean, std, input_size, etc.).

After this, the rest of the code can ask for one object (PreprocessConfig) instead of guessing transforms.

Mental model: “Look at the model filename; open the JSON next to it; parse those numbers into a neat Python object we can reuse everywhere.”

3) Build transforms/dataloader from the config

File: mithridatium/data.py

Add (simple version for CIFAR-10):

from torch.utils.data import DataLoader, Subset
from torchvision import datasets, transforms
from .io import PreprocessConfig

def build_transform(pp: PreprocessConfig):
    t = []
    for op in pp.ops:
        if op.startswith("resize:"):
            size = int(op.split(":")[1])
            t.append(transforms.Resize(size))
    t += [transforms.ToTensor(), transforms.Normalize(pp.mean, pp.std)]
    return transforms.Compose(t)

def build_dataloader(dataset_name: str, split: str, pp: PreprocessConfig, batch_size: int = 128, root: str = "./data"):
    tf = build_transform(pp)
    if dataset_name.lower() == "cifar10":
        ds = datasets.CIFAR10(root, train=(split=="train"), download=True, transform=tf)
    else:
        raise NotImplementedError(f"Dataset not supported: {dataset_name}")
    return DataLoader(ds, batch_size=batch_size, shuffle=(split=="train"))

Goal: Turn a PreprocessConfig into real torchvision transforms and a DataLoader.

build_transform(pp): reads the ops list (e.g., "resize:32"), builds the matching torchvision transforms, then adds ToTensor() and Normalize(mean, std).

build_dataloader(dataset_name, split, pp, ...): uses build_transform(pp) and creates a torch.utils.data.DataLoader for the named dataset (CIFAR-10 shown).

Mental model: “Take the config, build the image prep steps, and give me a PyTorch DataLoader ready to use.”

4) Example sidecar JSON (data-side)

File: models/resnet18_bd.json

Add:

{
  "preprocess": {
    "input_size": [32, 32],
    "channels_first": true,
    "value_range": [0.0, 1.0],
    "mean": [0.4914, 0.4822, 0.4465],
    "std":  [0.2023, 0.1994, 0.2010],
    "ops":  ["resize:32"]
  },
  "dataset": "cifar10"
}

Goal: Show what the sidecar looks like. It holds the exact preprocessing choices (size, mean/std, etc.) and the dataset name.

5) Use it in scripts/evaluator 

File: scripts/check_evaluator.py 

Change:

from mithridatium.modelcard import load_preprocess_config
from mithridatium.data import build_dataloader

pp = load_preprocess_config(args.model)
test_loader = build_dataloader("cifar10", "test", pp, batch_size=args.batch_size)

Goal: Replace ad-hoc transforms with the new standardized path.

Load the config from the sidecar (load_preprocess_config).

Build the dataloader from that config (build_dataloader).

Pass that dataloader to your evaluator/MMBD (their code doesn’t need to know the details).

6) Tests

File: tests/test_preprocess_config.py

Add basic checks:

Loading models/resnet18_bd.json returns a PreprocessConfig.

build_dataloader("cifar10","test", pp) returns a working loader.

Goal: Prove the plumbing works.

Load the JSON → get a PreprocessConfig.

Use it to build a dataloader and iterate one batch (just to confirm shapes flow end-to-end).

## Additional Context

What is a sidecar file?
A small JSON that sits next to the model (same name, .json extension) and tells the code how to preprocess inputs.
Example: model = models/resnet18_bd.pth → sidecar = models/resnet18_bd.json.

Why are we doing this?

- Avoids hardcoding CIFAR-10 transforms inside detectors.

- Makes the same preprocessing available to evaluator, detectors, and CLI.

- Update one JSON instead of digging through code.
